### PR TITLE
Add test bodies to test run results

### LINKED ISSUE #24: Left to do for version 2 compatibiltiy
1.  [x] capture debugging stdout
1. [x] associate the test body with the tests run.

### PR DESCRIPTION
**Pull request by glennj**
The purpose is to parse the test file to provide the code executed for each test as a result of the runner.

Closes #24

### CODE REVIEW

File: bin/run.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
-#!/bin/bash
+#!/usr/bin/env bash
 
-set -o errexit
-set -o nounset
+#set -o errexit
+#set -o nounset
```
Comment (IsaacG) [inline bin/run.sh]: If we don't want to use these, just drop them.

File: bin/run.sh [hunk 2]
```diff
@@ -77,15 +79,54 @@ run_tests() {
     # test scripts may be (old) xxxx_test.sh
     [[ -f "$test_file" ]] || test_file="${slug//-/_}_test.sh"
 
-    sed -i 's/load bats-extra.bash/load bats-extra/' "$test_file" || true
```
Comment (IsaacG) [inline bin/run.sh:80]: Is there value to swapping out sed for perl?

File: bin/run.sh [hunk 3]
```diff
@@ -77,15 +79,54 @@ run_tests() {
...
+            out)
+                if [[ $line =~ $start_test_re ]]; then
+                    name=${BASH_REMATCH[1]}
+                    body=()
+                    state="in"
+                fi
+                ;;
+            in)
+                if [[ $line =~ $end_test_re ]]; then
+                    test_bodies["$name"]=$(join $'\n' "${body[@]}")
```
Comment (IsaacG) [inline bin/run.sh]:
```suggestion
                    test_bodies["$name"]=$(printf '%s\n' "${body[@]}")
```

File: bin/run.sh [hunk 4]
```diff
@@ -77,15 +79,54 @@ run_tests() {
...
+    local name line
+    local state="out"
+    local body=() 
+    test_bodies=()
+
+    local start_test_re='^@test ['\''"](.+)['\''"] \{[[:blank:]]*$'
+    local end_test_re='^\}[[:blank:]]*$'
+
+    while IFS= read -r line; do
+        case $state in
```
Comment (IsaacG) [inline bin/run.sh]: Always quote all variable expansions.

File: bin/run.sh [hunk 5]
```diff
@@ -119,8 +161,12 @@ build_report() {
             error "$output_file" "$json_result_file"; return 1
         fi
 
+        [[ -v "test_bodies[$test_name]" ]] &&
+            test_body=${test_bodies[$test_name]} ||
+            test_body=""
```
Comment (IsaacG) [inline bin/run.sh]:
```suggestion
        test_body=${test_bodies[$test_name]:-}
```

File: bin/run.sh [hunk 6]
```diff
@@ -77,15 +79,54 @@ run_tests() {
...
+}
+
+get_test_bodies() {
+    local test_file=$1
+    local name line
+    local state="out"
+    local body=() 
+    test_bodies=()
+
+    local start_test_re='^@test ['\''"](.+)['\''"] \{[[:blank:]]*$'
```
Comment (IsaacG) [inline bin/run.sh:99]: Is this JSON parsing? Can we use `jq` or `awk` for parsing?

File: bin/run.sh (hunk 2 above)
Comment (glennj) [reply bin/run.sh:80]: I'm on a mac where the BSD sed's -i works differently from the GNU sed's -i, and I got tired of switching back and forth. There's no impactful performance hit, and perl is more porable.

File: bin/run.sh (hunk 6 above)
Comment (glennj) [reply bin/run.sh:99]: This is parsing the bats test file.

File: bin/run.sh (hunk 3 above)
Comment (glennj) [reply bin/run.sh]: Good suggestion.

File: bin/run.sh (hunk 4 above)
Comment (glennj) [reply bin/run.sh]: word splitting and glob expansion don't happen for the case var, but will do.

File: bin/run.sh (hunk 2 above)
Comment (IsaacG) [reply bin/run.sh:80]: That's annoying :( Oh well.

File: bin/run.sh (hunk 4 above)
Comment (IsaacG) [reply bin/run.sh]: It's standard "good style" advice. We should practice what we preach.

File: bin/run.sh (hunk 6 above)
Comment (IsaacG) [reply bin/run.sh:99]: If you want to parse in `awk` then read null-delineated data...
```
while IFS='' read -d '' name; read -d '' body; do printf '[%s][%s]\n' "$name" "$body"; done < <(
    awk '
        $0 ~ /^}$/ {i=0; printf "\00"}
        i {print}
        $1 == "@test" && $(NF) == "{" {
            sub(/^@test "/, "")
            sub(/" {$/, "")
            print $0, "\00"
            i=1
    } ' file.bats
)
```

Not sure if you find that harder or easier to read.

**Review by IsaacG — APPROVED** [review]

**Review by ErikSchierboom — APPROVED** [review]
LGTM!

My only nit would be to perhaps strip the two leading spaces from each line in the `test_code` field.

File: bin/run.sh (hunk 6 above)
Comment (glennj) [reply bin/run.sh:99]: I think I'm satisfied parsing the file in bash. I think the code is pretty clear.

**Pull request merged**